double callback and connection timeouts

after merging #70 and #72, we are now getting timeouts and double callbacks on connections.

```
$ make test
NODE_PATH=lib/ node_modules/whiskey/bin/whiskey --real-time --scope-leaks --tests "test/cql2.js test/cql3.js test/thriftHostPool.js test/thrift.js test/units.js"
/home/travis/builds/simplereach/helenus/test/cql2.js
set_cql_version not supported, unsetting cqlVersion
whiskey.common: test.finish in [bold]test a bad connection will return an error[/bold] has been called twice, possible double callback in your code! {"name":"test a bad connection will return an error"}
  setUp                                                                      [OK]
  test a bad connection will return an error                                 [OK]
  test cql create keyspace                                                   [OK]
  test cql use keyspace                                                      [OK]
  test cql create column family                                              [OK]
  test cql create counter column family                                      [OK]
  test cql create reversed column family                                     [OK]
  test cql update                                                            [OK]
  test cql update reversed                                                   [OK]
  test cql update with no callback                                           [OK]
  test cql incr                                                              [OK]
  test cql select                                                            [OK]
  test cql select reversed                                                   [OK]
  test cql select *                                                          [OK]
  test cql select counter                                                    [OK]
  test cql incr and select                                                   [OK]
  test cql select with bad user input                                        [OK]
  test cql count                                                             [OK]
  test cql error                                                             [OK]
  test cql count with gzip                                                   [OK]
  test cql delete                                                            [OK]
  test cql drop column family                                                [OK]
  test cql drop keyspace                                                     [OK]
  test too many cql params                                                   [OK]
  test too few cql params                                                    [OK]
  tearDown                                                                   [OK]
  timeout                                                               [TIMEOUT]
/home/travis/builds/simplereach/helenus/test/cql3.js
set_cql_version not supported, skipping CQL3 tests
/home/travis/builds/simplereach/helenus/test/thriftHostPool.js
whiskey.common: test.finish in [bold]test bad pool connect[/bold] has been called twice, possible double callback in your code! {"name":"test bad pool connect"}
  setUp                                                                      [OK]
  test pool.connect without keyspace                                         [OK]
  test pool.close without keyspace                                           [OK]
  test pool.connect with keyspace                                            [OK]
  test bad pool connect                                                      [OK]
  test pool.createKeyspace                                                   [OK]
  test pool.use                                                              [OK]
  test keyspace.createColumnFamily                                           [OK]
  test keyspace.createColumnFamily with composite type                       [OK]
  test keyspace.createSuperColumnFamily                                      [OK]
  test keyspace.createCounterFamily                                          [OK]
  test keyspace.get                                                          [OK]
  test keyspace.get composite                                                [OK]
  test keyspace.get supercolumn                                              [OK]
  test keyspace.get counter                                                  [OK]
  test keyspace.get from cache                                               [OK]
  test keyspace.get from index                                               [OK]
  test keyspace.get invalid cf                                               [OK]
  test standard cf.insert                                                    [OK]
  test standard cf.insert with custom CL                                     [OK]
  test standard cf.insert into composite cf                                  [OK]
  test counter cf.incr                                                       [OK]
  test standard cf.get                                                       [OK]
  test standard cf.get for composite column family                           [OK]
  test standard cf.get with options                                          [OK]
  test standard cf.get with custom CL                                        [OK]
  test standard cf.get with columns names                                    [OK]
  test composite cf.get with columns names                                   [OK]
  test counter cf.get with columns names                                     [OK]
  test standard cf.get with error                                            [OK]
  test standard cf with Index                                                [OK]
  test standard cf with BytestType                                           [OK]
  test standard cf with LongType                                             [OK]
  test standard cf with IntegerType                                          [OK]
  test standard cf with UTF8Type                                             [OK]
  test standard cf with AsciiType                                            [OK]
  test standard cf with LexicalUUIDType                                      [OK]
  test standard cf with TimeUUIDType                                         [OK]
  test standard cf with FloatType                                            [OK]
  test standard cf with DoubleType                                           [OK]
  test standard cf with DateType                                             [OK]
  test standard cf with BooleanType                                          [OK]
  test standard cf with UUIDType                                             [OK]
  test row.nameSlice                                                         [OK]
  test row.slice                                                             [OK]
  test row.toString and row.inspect                                          [OK]
  test row.forEach                                                           [OK]
  test standard cf remove column                                             [OK]
  test composite cf remove column                                            [OK]
  test standard cf remove row                                                [OK]
  test standard cf truncate                                                  [OK]
  test keyspace.dropColumnFamily                                             [OK]
  test pool.dropKeyspace                                                     [OK]
  test pool.close                                                            [OK]
  tearDown                                                                   [OK]
  timeout                                                               [TIMEOUT]
/home/travis/builds/simplereach/helenus/test/thrift.js
whiskey.common: test.finish in [bold]test bad pool connect[/bold] has been called twice, possible double callback in your code! {"name":"test bad pool connect"}
  setUp                                                                      [OK]
  test pool.connect without keyspace                                         [OK]
  test pool.close without keyspace                                           [OK]
  test pool.connect with keyspace                                            [OK]
  test bad pool connect                                                      [OK]
  test pool.createKeyspace                                                   [OK]
  test pool.use                                                              [OK]
  test keyspace.createColumnFamily                                           [OK]
  test keyspace.createColumnFamily with composite type                       [OK]
  test keyspace.createSuperColumnFamily                                      [OK]
  test keyspace.createCounterFamily                                          [OK]
  test keyspace.get                                                          [OK]
  test keyspace.get composite                                                [OK]
  test keyspace.get supercolumn                                              [OK]
  test keyspace.get counter                                                  [OK]
  test keyspace.get from cache                                               [OK]
  test keyspace.get from index                                               [OK]
  test keyspace.get invalid cf                                               [OK]
  test standard cf.insert                                                    [OK]
  test standard cf.insert with custom CL                                     [OK]
  test standard cf.insert into composite cf                                  [OK]
  test counter cf.incr                                                       [OK]
  test standard cf.get                                                       [OK]
  test standard cf.get for composite column family                           [OK]
  test standard cf.get with options                                          [OK]
  test standard cf.get with custom CL                                        [OK]
  test standard cf.get with columns names                                    [OK]
  test composite cf.get with columns names                                   [OK]
  test counter cf.get with columns names                                     [OK]
  test standard cf.get with error                                            [OK]
  test standard cf with Index                                                [OK]
  test standard cf with BytestType                                           [OK]
  test standard cf with LongType                                             [OK]
  test standard cf with IntegerType                                          [OK]
  test standard cf with UTF8Type                                             [OK]
  test standard cf with AsciiType                                            [OK]
  test standard cf with LexicalUUIDType                                      [OK]
  test standard cf with TimeUUIDType                                         [OK]
  test standard cf with FloatType                                            [OK]
  test standard cf with DoubleType                                           [OK]
  test standard cf with DateType                                             [OK]
  test standard cf with BooleanType                                          [OK]
  test standard cf with UUIDType                                             [OK]
  test row.nameSlice                                                         [OK]
  test row.slice                                                             [OK]
  test row.toString and row.inspect                                          [OK]
  test row.forEach                                                           [OK]
  test standard cf remove column                                             [OK]
  test composite cf remove column                                            [OK]
  test standard cf remove row                                                [OK]
  test standard cf truncate                                                  [OK]
  test keyspace.dropColumnFamily                                             [OK]
  test pool.dropKeyspace                                                     [OK]
  test pool.close                                                            [OK]
  tearDown                                                                   [OK]
  timeout                                                               [TIMEOUT]
/home/travis/builds/simplereach/helenus/test/units.js
  setUp                                                                      [OK]
  test uuid library                                                          [OK]
---------------------------------------------------------------------------------
Ran 138 tests in 45.345s

Successes: 138
Failures: 0
Timeouts: 3
Skipped: 0

FAILED

Leaked Variables

  cql2.js: errno
  thriftHostPool.js: errno
  thrift.js: errno
  units.js: no leaks detected
make: *** [test] Error 3

Done. Build script exited with: 1
```

--- a/lib/pool.js
+++ b/lib/pool.js
@@ -69,7 +69,7 @@ util.inherits(Pool, process.EventEmitter);
  * @param {Function} callback The callback to invoke when all connections have been made
  */
 Pool.prototype.connect = function(callback){
-  var i = 0, finished = 0, self = this, 
+  var i = 0, finished = 0, self = this,
       len = this.hosts.length * this.hostPoolSize,
       connected = 0;
 
@@ -82,12 +82,12 @@ Pool.prototype.connect = function(callback){
       connected += 1;
       self.clients.push(connection);
     }
-     
+
     if(finished === len){
       if(self.clients.length === 0){
-        replyNotAvailable(callback);
+        return replyNotAvailable(callback);
       }
-      
+
       //set the keyspaces connection to be the pool
       if(keyspace){
         keyspace.connection = self;

--- a/lib/pool.js
+++ b/lib/pool.js
@@ -316,12 +316,17 @@ Pool.prototype.close = function(){
 
   clearInterval(this.retryInterval);
 
+  if (len === 0){
+    this.emit('close');
+  }
+
   function closed(){
     j += 1;
     if(j === len){
       self.emit('close');
     }
   }
+
   for(; i < len; i += 1){
     this.clients[i].on('close', closed);
     this.clients[i].close();

--- a/lib/pool.js
+++ b/lib/pool.js
@@ -84,17 +84,16 @@ Pool.prototype.connect = function(callback){
     }
 
     if(finished === len){
-      if(self.clients.length === 0){
-        return replyNotAvailable(callback);
-      }
-
       //set the keyspaces connection to be the pool
       if(keyspace){
         keyspace.connection = self;
       }
       //we only want to callback once, after we get the final connection
-      callback(null, keyspace);
-
+      if(self.clients.length === 0){
+        replyNotAvailable(callback);
+      } else {
+        callback(null, keyspace);
+      }
       //now that we have a connection, lets start monitoring
       self.monitorConnections();
     }

--- a/test/cql2.js
+++ b/test/cql2.js
@@ -33,6 +33,7 @@ module.exports = {
         assert.isDefined(err);
         badConn.cql(config['create_ks#cql'], function(err, res){
            assert.isDefined(err);
+           badConn.close();
            test.finish();
          });
      });

--- a/test/cql3.js
+++ b/test/cql3.js
@@ -60,6 +60,7 @@ module.exports = {
         assert.isDefined(err);
         badConn.cql(config['create_ks#cql'], function(err, res){
            assert.isDefined(err);
+           badConn.close();
            test.finish();
          });
      });

--- a/test/thrift.js
+++ b/test/thrift.js
@@ -47,6 +47,7 @@ module.exports = {
           assert.isDefined(err);
           badConn.dropKeyspace(config.keyspace, function(err){
             assert.isDefined(err);
+            badConn.close();
             test.finish();
           });
        });

